internal/observability/event/sink_file.go

```diff
@@ -189,9 +189,23 @@ func (s *FileSink) open() error {
 }
 
 // log writes the buffer to the file.
-// It acquires a lock on the file to do this.
-func (s *FileSink) log(data []byte) error {
-	s.fileLock.Lock()
+// NOTE: We attempt to acquire a lock on the file in order to write, but will
+// yield if the context is 'done'.
+func (s *FileSink) log(ctx context.Context, data []byte) error {
+	// Don't just wait for the lock, ignorant of the context as we could end up
+	// with a lot of entries queuing up for access to the sink. If our context gets
+	// cancelled then we should just give up and not continue to wait for our turn.
+contextDoneOrLock:
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			if s.fileLock.TryLock() {
```

Comment:
Should there be a small sleep in an `else` case here? I worry that something could get spinning in here up to 100% CPU. If the lock's not ready then this thread is super busy spinning through this loop non-stop. We had a similar bug recently in Agent/Proxy leading to high CPU usage.

Reply:
This same question is true for the similar bit of code in `internal/observability/event/sink_socket.go`

Reply:
Yeah, it's a good point. I've pushed up some changes to just let us queue for the lock (letting Go figure it out) then check the context straight away when we get the lock (potentially just releasing the lock if we're 'done').